File: frontend/src/types/competency.ts
export interface CompetencyItem {
  id: string;
  name: string;
  domain: string;
  description?: string;
  created_at: string;
}

export interface CompetencyTaxonomyResponse {
  total: number;
  competencies: CompetencyItem[];
}

export interface TraineeCompetencyItem {
  id: string;
  competency_id: string;
  name: string;
  domain: string;
  proficiency_level: number;
  level_percentage: number;
  last_evaluated_at: string;
}

export interface TraineeCompetencyMatrixResponse {
  user_id: string;
  user_name: string;
  competencies: TraineeCompetencyItem[];
}

export interface CompetencyGapItem {
  competency: string;
  competency_id?: string;
  required: number;
  current: number;
  gap: number;
  is_met: boolean;
}

export interface SkillGapAnalysisResponse {
  target_role: string;
  target_role_title: string;
  overall_readiness_percentage: number;
  total_gap_magnitude: number;
  gaps: CompetencyGapItem[];
}

export interface CourseRecommendationItem {
  course_id: string;
  title: string;
  code: string;
  match_score: number;
  match_percentage: number;
  rationale: string;
  thumbnail_url?: string;
  imparted_competencies: Array<{
    competency: string;
    yield_level: number;
  }>;
}

export interface CourseCompetencyItem {
  competency_id: string;
  competency_name: string;
  domain: string;
  yield_level: number;
}

export interface TrainerMatchItem {
  trainer_id: string;
  full_name: string;
  email: string;
  station_code?: string;
  years_of_experience: number;
  satisfaction_rating: number;
  expertise_similarity: number;
  experience_score: number;
  satisfaction_score: number;
  composite_score: number;
  match_percentage: number;
  matched_expertise?: string;
  rationale: string;
}

export interface TrainerMatchResponse {
  subject: string;
  total_matched: number;
  trainers: TrainerMatchItem[];
}

export interface RoleBenchmarkResponse {
  role_key: string;
  title: string;
  description: string;
  requirements: Record<string, number>;
}

File: backend/app/schemas/competency.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime


class CompetencyBase(BaseModel):
    name: str = Field(..., max_length=150, description="Standardized competency name")
    domain: str = Field(..., max_length=100, description="Meteorological operational domain")
    description: Optional[str] = None
    criticality_weight: float = Field(default=1.0, gt=0.0)


class CompetencyCreate(CompetencyBase):
    pass


class CompetencyResponse(CompetencyBase):
    id: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CompetencyTaxonomyResponse(BaseModel):
    total: int
    competencies: List[CompetencyResponse]


class TraineeCompetencyItem(BaseModel):
    id: str
    competency_id: str
    name: str
    domain: str
    proficiency_level: float = Field(..., ge=0.0, le=1.0)
    level_percentage: float = Field(..., ge=0.0, le=100.0)
    last_evaluated_at: datetime


class TraineeCompetencyMatrixResponse(BaseModel):
    user_id: str
    user_name: str
    competencies: List[TraineeCompetencyItem]


class TraineeCompetencyUpdateRequest(BaseModel):
    competency_id: str
    proficiency_level: float = Field(..., ge=0.0, le=1.0)


class CompetencyGapItem(BaseModel):
    competency: str
    competency_id: Optional[str] = None
    required: float
    current: float
    gap: float
    weighted_gap: float
    is_met: bool


class SkillGapAnalysisResponse(BaseModel):
    target_role: str
    target_role_title: str
    overall_readiness_percentage: float
    total_gap_magnitude: float
    gaps: List[CompetencyGapItem]


class CourseCompetencyItem(BaseModel):
    competency_id: str
    competency_name: str
    domain: str
    yield_level: float = Field(..., ge=0.0, le=1.0)


class CourseCompetencyMapRequest(BaseModel):
    competency_id: str
    yield_level: float = Field(default=0.50, ge=0.0, le=1.0)


class AssessmentCompetencyMapRequest(BaseModel):
    competency_id: str


class CourseRecommendationItem(BaseModel):
    course_id: str
    title: str
    code: str
    match_score: float = Field(..., ge=0.0, le=1.0)
    match_percentage: float = Field(..., ge=0.0, le=100.0)
    rationale: str
    thumbnail_url: Optional[str] = None
    imparted_competencies: List[Dict[str, Any]] = []


class TrainerMatchRequest(BaseModel):
    subject: str = Field(..., min_length=2, description="Target subject/topic to match trainer for")
    minimum_experience_years: Optional[float] = Field(default=0.0, ge=0.0)
    competency_id: Optional[str] = None


class TrainerMatchItem(BaseModel):
    trainer_id: str
    full_name: str
    email: str
    station_code: Optional[str] = None
    years_of_experience: float
    satisfaction_rating: float
    expertise_similarity: float
    experience_score: float
    satisfaction_score: float
    composite_score: float
    match_percentage: float
    matched_expertise: Optional[str] = None
    availability_confirmed: bool
    rationale: str


class TrainerMatchResponse(BaseModel):
    subject: str
    total_matched: int
    trainers: List[TrainerMatchItem]


class RoleBenchmarkResponse(BaseModel):
    role_key: str
    title: str
    description: str
    requirements: Dict[str, float]
